(PDB-3275) Add generative command tests

mullr opened the pull request.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test PASSed

puppetlabs-jenkins commented:
Test PASSed

puppetcla commented:
CLA signed by all contributors.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
This is going to need some squashing at the end of the data, but I think I've ferreted out most of the failures. Fixed two real bugs in the process, too.

puppetlabs-jenkins commented:
Test FAILed

senior commented on test/puppetlabs/puppetdb/generative/submit_command.clj:
```diff
@@ -0,0 +1,151 @@
+                      "2016-12-19T23:39:00.000Z"})
+
+(defmacro with-catalog-overrides [body]
+  `(with-redefs [pdb-gen/certname #(gen/elements ["a.com" "b.com" "c.com"])
```
Talked with @mullr some about this. I think it's really important that we have this kind of override capability. I'd like to see it plumbed through as first class. Sounds like having that gets us a little closer to what the spec version will look like as well which is a nice bonus

mullr commented:
@senior So... 

I... um... did a thing. 

Please understand, I've had no outlet for this stuff for so long....

mullr commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -1,109 +1,180 @@
             [com.gfredericks.test.chuck.generators :as chuck-gen]
-            [clj-time.core :as time :refer [now]]))
+            [clj-time.core :as time :refer [now]]
+            [clojure.tools.logging :as log]
```
cider added this for me, shouldn't be here.

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
@senior Alternatively, I could do it with a dynamic var :wink:

mullr commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -0,0 +1,297 @@
+;;; catalog generator
+
+(defn certname [] (string+))
+(defn environment [] (string+))
```
I think the main generator definition here should be a the definition of what the environment /can/, be, i.e. keep it close to what a spec would be. But I can see doing that in the overrides. Not sure if test.check knows how to do that kind of thing, but I'm sure we could teach it.

mullr commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -0,0 +1,297 @@
+
+(defn resource-type [] (string+))
+(defn resource-title [] (string+))
+(defn resource_spec []
```
oh, that shouldn't be an underscore.

mullr commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -0,0 +1,297 @@
+                    "subscription-of"])))
+
+(defn edge []
+  (gen/hash-map :source (source)
```
I think this one is actually stale; see the resource-tree generator and resource-tree-to-edge-list function for the real thing. It only makes trees, but at least it's sort of real.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
@senior I think this is pretty baked now. I cleaned up the patchset and improved the docs on the overridable generator code, and added a few meta-tests for good measure. Still not 100% convinced it's what we should do, but I now at least think the idea is sound.

wkalt commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -0,0 +1,285 @@
+                 "before"
+                 "subscription-of"]))
+
+
```
extra spaces

mullr replied:
:chec

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test PASSed

puppetlabs-jenkins commented:
Test PASSed

senior commented:
@mullr Looks like the zip-test stuff is still failing, have you been able to look into that?

senior commented on test/puppetlabs/puppetdb/generative/spec.clj:
```diff
@@ -0,0 +1,119 @@
+(comment
```
It seems like we should remove this as there's no way for us to make sure the code still works, maybe it can be squirreled away in a branch?

mullr replied:
I'll put it in a gist

mullr replied:
https://gist.github.com/mullr/e596ec9efdd7c89aba65026aea705140

mullr commented:
@senior fixed the zip-test issues

puppetlabs-jenkins commented:
Test PASSed

senior commented:
@mullr you've got some failing tests on travis, out of memory. Maybe it's as easy as dialing down some of the generators?

mullr commented:
@senior ah forgot about that. yes, it should be. I think the generators get a little bit rowdy up towards the high iterations.

puppetlabs-jenkins commented:
Test PASSed

senior commented on test/puppetlabs/puppetdb/generative/overrideable_generators_test.clj:
```diff
@@ -0,0 +1,29 @@
+(ns puppetlabs.puppetdb.generative.overrideable-generators-test
+  (:require [clojure.test :as t]
```
Unless there's a solid reason for this, I'd prefer to keeps this consistent with all of our other tests and just :refer :all

senior commented on test/puppetlabs/puppetdb/generative/submit_command.clj:
```diff
@@ -0,0 +1,150 @@
+
+(def overrides {:puppet/certname (gen/elements ["a.com" "b.com" "c.com"])
+                :catalog/version (gen/return "9")
+                :puppet/environment (gen/return "production")
```
Seems like we should have a few of these

senior commented on test/puppetlabs/puppetdb/generative/submit_command.clj:
```diff
@@ -0,0 +1,150 @@
+             (gen/override overrides)
+             gen/no-shrink
+             gen/convert)]
+   (print "-") (flush)
```
I think this might be the first time we're intentionally printing things in a test. It seems useful when working on the tests locally, but it seems like it would just be noise in CI. Not sure what the best way to accommodate both is. Maybe log/trace or something? Thoughts @rbrw?

mullr replied:
I put these in initially to be able to tell the difference between test.check going into shrink mode (adding thousands of iterations) and a generator being completely out to lunch. Since everything is behind gen/no-shrink now, I think I can just get rid of it.

senior commented on test/puppetlabs/puppetdb/generative/generators.clj:
```diff
@@ -0,0 +1,283 @@
+;;; factset generator
+
+(defgen :fact/name (string+))
+(defgen :fact/value json-value)
```
Will this only generate vectors and maps? Or will it generate strings/integers as well (not inside a tree structure)

mullr replied:
I *think* it'll give you plain values sometimes. Checking...

mullr replied:
wow, it acutally doesn't. I didn't expect that.

senior replied:
👍

mullr commented on test/puppetlabs/puppetdb/generative/overridable_generators.clj:
```diff
@@ -0,0 +1,135 @@
+                        (recur gen-val overrides)
+                        (throw (ex-info (str "Generator " gen " not registered") {:generator-key gen})))
+       (gen/generator? gen) gen
+       :default (gen overrides)))))
```
You mean `(gen gens)`? No, I don't think so. I mean, I guess we could, but there's no point. In this case, 'gen' is an overridable genearator, i.e. a function of overrides -> generator. Each overridable generator calls `lookup` on its arguments, passing down overrides. So eventually it's going to bottom out in this `lookup` function again, having either a regular generator or a keyword. 

Which is to say, if we passed down `gens`, we'd end up merging in `@generators` twice. 

I do think moving the `let` into the `(keyword?)` block would be better though.

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
rebased on top of latest master

puppetlabs-jenkins commented:
Test PASSed
